Set target-abi module flag for RISC-V targets

kxxt opened the pull request:
Fixes cross-language LTO on RISC-V targets (Fixes #121924)

kxxt committed:
Set target-abi module flag for RISC-V targets

Fixes cross-language LTO on RISC-V targets (Fixes #121924)

jieyouxu commented:
Relevant LLVM patch: https://github.com/llvm/llvm-project/commit/227496dc09cf46df233aad041d6dc6113822e4bb

Some remarks (I'm not very familiar with LLVM + RISC-V + LTO):
- This will definitetly need a regression test to make sure we don't accidentally break things in the future, I believe your test case could be written as a ran-make test for the specific riscv target.
- Can you leave a FIXME for emitting RISC-V-specific target-abi metadata linking to the relevant LLVM issue? AFAICT this is still considered a workaround?
- Should we unconditionally emit target-abi metadata?

jieyouxu commented:
Actually I just did a double take: are you sure this isn't a lld problem? What happens if your reproducer makefile specifies `-Wl,-plugin-opt=-target-abi=lp64d`?

kxxt commented:
Hi @jieyouxu, Thanks for your quick review.

> Would a current workaround for the reproducer makefile be to specify `-Wl,-plugin-opt=-target-abi=lp64d`? For rustc it's something like `-Clink-args="-target-abi=lp64d"`

Yes, [that will work](https://github.com/kxxt/riscv-rust-cross-language-lto-reproducer/commit/ad619e4796bc04f6aaa1a0ff97dfe644585c5a62), but I still think this should be fixed in rustc.

> Should we unconditionally emit target-abi metadata?

I am not familiar with rustc or llvm. But from the code it does look like clang is emitting this metadata unconditionally: https://github.com/llvm/llvm-project/blob/llvmorg-19-init/clang/lib/CodeGen/CodeGenModule.cpp#L1057-L1062

> Can you leave a FIXME for emitting RISC-V-specific target-abi metadata linking to the relevant LLVM issue? AFAICT this is still considered a workaround?

I will do it.

> This will definitetly need a regression test to make sure we don't accidentally break things in the future, I believe your test case could be written as a run-make test for the specific riscv target to make sure it works across LTO, and maybe a codegen(?) test checking that we emit the target-abi metadata for the riscv target.

I will try to add tests.

jieyouxu commented:
I personally think that emitting target-abi does make sense. But I'm far from familiar with the LLVM side of things to judge if this approach is what makes the most sense given the context, so I'll leave this to reviewers more well-versed in the LLVM side of things to give the final reviews.

kxxt commented:
I have added a run-make test and a codegen test and fixed a bug that empty target-abi got emitted for targets that have llvm_abiname defined as an empty string.

jieyouxu reviewed (changes requested):
Thanks for adding the tests. I left some reviews for the tests themselves. For the LLVM wrapper and LLVM codegen changes, another reviewer will look at them.

kxxt commented on tests/run-make/cross-lang-lto-riscv-abi/Makefile:
```diff
@@ -0,0 +1,24 @@
+# needs-matching-clang
```
> Please try to [write this test in rmake.rs](https://github.com/rust-lang/rust/issues/121876), we're trying to get rid of Makefiles in our run-make test suite.

Thanks! I don't know that beforehand.

jieyouxu replied:
No worries, the migration is still on-going and it's easy to following existing Makefiles and write new ones.

kxxt commented on tests/codegen/riscv-target-abi.rs:
```diff
@@ -0,0 +1,20 @@
+
+//@[riscv64gc] compile-flags: --target=riscv64gc-unknown-linux-gnu
+//@[riscv64gc] needs-llvm-components: riscv
+// riscv64gc: !{i32 {{1}}, !"target-abi", !"lp64d"}
```
> I am almost certain `// riscv64gc: !{i32 {{1}}, !"target-abi", !"lp64d"}` doesn't check anything.

It works. If I change "lp64d" to "lp64" the test would fail.

jieyouxu replied:
https://github.com/rust-lang/rust/blob/75fd074338801fba74a8cf7f8c48c5c5be362d08/src/tools/compiletest/src/runtest.rs#L2883-L2892

You're right... that is... quite surprising 😓

kxxt commented on tests/run-make/cross-lang-lto-riscv-abi/Makefile:
```diff
@@ -0,0 +1,24 @@
+@echo "Testing target $(1)"
+$(RUSTC) --target $(1) -Clinker-plugin-lto=on -Cpanic=abort --crate-type=rlib -o $(TMPDIR)/libriscv-xlto.a ./riscv-xlto.rs
+$(CLANG) -target $(2) -march=$(3) -mabi=$(4) -flto=thin -fuse-ld=lld -L $(TMPDIR) -lriscv-xlto -nostdlib -o $(TMPDIR)/riscv-xlto ./cstart.c
+file $(TMPDIR)/riscv-xlto | $(CGREP) "$(5)"
```
> I would encourage you to try to [edit PR CI](https://rustc-dev-guide.rust-lang.org/contributing.html?highlight=PR%20CI#ci) and try to run the rmake.rs version of this test with the riscv runner.

Could you share more details about how to edit PR CI? The linked doc doesn't tell how to do it.

> with the riscv runner.

Does this mean that the test runs on a riscv64 host? This test could run on regular x86_64 host as well.

jieyouxu replied:
Ah wait, this is just checking if the library contains the string ABI? Sorry I thought for some reason you were running some executable.

jieyouxu replied:
Ah, but:
> needs-matching-clang — ignores if the version of clang does not match the LLVM version of rustc. These tests are always ignored unless a special environment variable is set (which is only done in one CI job).
> -- <https://rustc-dev-guide.rust-lang.org/tests/headers.html?highlight=needs-matching-clang#ignoring-tests>

you may want to still edit PR CI to try run this specific CI job (I'll go ask around which one this is, lol) so if it fails, it doesn't fail in the full CI build.

See [Testing with CI](https://rustc-dev-guide.rust-lang.org/tests/ci.html?highlight=ci.yml#using-ci-to-test) for how to edit PR CI.

kxxt replied:
> needs-matching-clang 
Actually I am testing locally with a non-matching-clang (clang16) and it works as well. But other cross-language LTO run-makes have this condition so I added it as well.

jieyouxu replied:
Then it's probably fine.

kxxt committed:
Convert tests/run-make/cross-lang-lto-riscv-abi to rmake

kxxt commented on src/tools/tidy/src/allowed_run_make_makefiles.txt:
```diff
@@ -36,6 +36,7 @@ run-make/crate-hash-rustc-version/Makefile
 run-make/crate-name-priority/Makefile
 run-make/cross-lang-lto-clang/Makefile
 run-make/cross-lang-lto-pgo-smoketest/Makefile
+run-make/cross-lang-lto-riscv-abi/Makefile
```
Now it's converted to a rmake test.

nikic commented on compiler/rustc_codegen_llvm/src/context.rs:
```diff
@@ -326,6 +327,23 @@ pub unsafe fn create_module<'ll>(
+    // correctly setting target-abi for the LTO object
+    // FIXME: https://github.com/llvm/llvm-project/issues/50591
+    // If llvm_abiname is empty, emit nothing.
+    if matches!(sess.target.arch.as_ref(), "riscv32" | "riscv64" | "riscv128")
```
We don't have a riscv128 target. If you want to be forward compatible to that degree, I'd make this a starts_with.

kxxt replied:
Fixed. I removed `riscv128`. I am intentionally avoiding `starts_with` because another architecture might(possibly not) appear in the future that starts with `riscv`.

nikic commented on tests/codegen/riscv-target-abi.rs:
```diff
@@ -0,0 +1,20 @@
+
+//@[riscv64gc] compile-flags: --target=riscv64gc-unknown-linux-gnu
+//@[riscv64gc] needs-llvm-components: riscv
+// riscv64gc: !{i32 {{1}}, !"target-abi", !"lp64d"}
```
```suggestion
// riscv64gc: !{i32 1, !"target-abi", !"lp64d"}
```
Why the `{{1}}`?

kxxt replied:
Fixed. I directly copied it from the output of `llvm-dis`.

nikic commented on tests/codegen/riscv-target-abi.rs:
```diff
@@ -0,0 +1,20 @@
+
+//@[riscv32imac] compile-flags: --target=riscv32imac-unknown-none-elf
+//@[riscv32imac] needs-llvm-components: riscv
+// riscv32imac-NOT: !{i32 {{1}}, !"target-abi", !""}
```
```suggestion
// riscv32imac-NOT: !"target-abi"
```
I'd be more liberal with the CHECK-NOT, otherwise there is a risk it will not trigger due to some minor difference in the output.

kxxt replied:
Fixed

nikic reviewed (commented):
The LLVM changes look reasonable to me.

DianQK commented on tests/run-make/cross-lang-lto-riscv-abi/rmake.rs:
```diff
@@ -0,0 +1,75 @@
+}
+
+fn main() {
+    check_target("riscv64gc-unknown-linux-gnu", "riscv64-linux-gnu", "rv64gc", "lp64d", true);
```
Can we add or just test the default values for `-mabi` and `-march`?

kxxt replied:
I think we can drop `-mabi` because probably clang could compute that from `-march`

DianQK replied:
I can reproduce the issue when dropping `-march`. I'm not sure if we also need this parameter. It might not matter.

dianqk reviewed (commented):
I think we can drop cffc0f5ce2470540578bf3d28a7cfd91f72a2166 now.

DianQK commented on tests/run-make/cross-lang-lto-riscv-abi/rmake.rs:
```diff
@@ -0,0 +1,75 @@
+    }
+    // Check that the built binary has correct float abi
+    let llvm_readobj =
+        PathBuf::from(env::var("LLVM_BIN_DIR").unwrap()).join(bin_name("llvm-readobj"));
```
@jieyouxu 
Just a question, unrelated to this PR: Is putting this type of code into `run_make_support` part of the plan?

jieyouxu replied:
If it feels like too much work, that is also okay. I'm happy to accept this PR as is without refactoring out the commands to helper structs in the support library, I can write a follow up PR to do the refactoring (there might be some API experimentation work).

jieyouxu replied:
I thought about it some more: I'm happy with accepting this PR as is. I'll follow up with a refactor PR so kxxt doesn't have to do the support library work (which feels like unnecessary busy work for the intent of this PR).

dianqk reviewed (approved):
I can reproduce the issue when dropping `-march`. I'm not sure if we also need this parameter. It might not matter.

For me, using `drop` instead of `revert` is better. This commit does not contribute to the history.

kxxt commented:
> I can reproduce the issue when dropping `-march`. I'm not sure if we also need this parameter. It might not matter.

I don't think we should drop `-march`, because the clang target triple doesn't specify the extensions(especially if the d ext is available, which should be what causing the default float abi to be different)

> For me, using `drop` instead of `revert` is better. This commit does not contribute to the history.

Should I keep all the micro commits or squash them?

DianQK commented:
> > For me, using `drop` instead of `revert` is better. This commit does not contribute to the history.
> 
> Should I keep all the micro commits or squash them?

As far as I know, the current rust repository has very liberal requirements for commits. I think it's fine to keep them as long as they are helpful for reviewing the commit history. :)

kxxt commented:
> > > For me, using `drop` instead of `revert` is better. This commit does not contribute to the history.
> > 
> > 
> > Should I keep all the micro commits or squash them?
> 
> As far as I know, the current rust repository has very liberal requirements for commits. I think it's fine to keep them as long as they are helpful for reviewing the commit history. :)

Thanks for your review! I have rebased the history into two commits.

yujincheng08 commented on compiler/rustc_codegen_llvm/src/context.rs:
```diff
@@ -326,6 +327,23 @@ pub unsafe fn create_module<'ll>(
+    if matches!(sess.target.arch.as_ref(), "riscv32" | "riscv64")
+        && !sess.target.options.llvm_abiname.is_empty()
+    {
+        let llvm_abiname = CString::new(sess.target.options.llvm_abiname.as_ref()).unwrap();
```
instead of creating a new `cstr`, I think we can follow other parts to pass `value` and `valueLen` to create a StringRef

kxxt replied:
Thanks for your review! This is clearly a better way.

yujincheng08 commented on compiler/rustc_codegen_llvm/src/context.rs:
```diff
@@ -326,6 +327,23 @@ pub unsafe fn create_module<'ll>(
+        llvm::LLVMRustAddModuleFlagString(
+            llmod,
+            llvm::LLVMModFlagBehavior::Error,
+            c"target-abi".as_ptr() as *const _,
+            llvm_abiname.as_ptr() as *const _,
```
```suggestion
            c"target-abi".as_ptr(),
            sess.target.llvm_abiname.as_ptr().cast(),
            sess.target.llvm_abiname.len(),
```

yujincheng08 commented on compiler/rustc_codegen_llvm/src/llvm/ffi.rs:
```diff
@@ -1801,12 +1801,20 @@ extern "C" {
     /// In order for Rust-C LTO to work, module flags must be compatible with Clang. What
     /// "compatible" means depends on the merge behaviors involved.
-    pub fn LLVMRustAddModuleFlag(
+    pub fn LLVMRustAddModuleFlagU32(
```
maybe we can keep its name so that the change can be smaller; consider only riscv needs to set a string flag.

kxxt replied:
I think it doesn't hurt because the current naming scheme is more consistent(method name with a type suffix. And we don't have to care about patch size.

kxxt committed:
Pass value and valueLen to create a StringRef

Instead of creating a cstring.

Co-authored-by: LoveSy <shana@zju.edu.cn>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8611497531/job/23598796476) [(plain)](https://github.com/rust-lang/rust/commit/544fc109c76cf36b17bb2dedd8bf38de050240e3/checks/23598796476/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.6s done
#16 DONE 33.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
     |
339  |           llvm::LLVMRustAddModuleFlagString(
     |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
343  |               sess.target.llvm_abiname.as_ptr().cast(),
344  | |             sess.target.llvm_abiname.len(),
     | |             ------------------------------
     | |_____________|____________________________|
     |               |                            help: remove the extra argument
```

</details>

yujincheng08 commented on compiler/rustc_codegen_llvm/src/llvm/ffi.rs:
```diff
@@ -1801,12 +1801,20 @@ extern "C" {
+        M: &Module,
+        merge_behavior: LLVMModFlagBehavior,
+        name: *const c_char,
+        value: *const c_char,
```
```suggestion
        value: *const c_char,
        valueLen: size_t,
```

yujincheng08 replied:
sorry, I forgot this 🤦

jieyouxu commented:
r? @jieyouxu

jieyouxu commented on tests/run-make/cross-lang-lto-riscv-abi/rmake.rs:
```diff
@@ -0,0 +1,74 @@
+        .linker_plugin_lto("on")
+        .run();
+    // C part
+    let clang = env::var("CLANG").unwrap();
```
Probably also convert `clang` into a wrapper struct in `run_make_support` so other tests can use a helper from the support library.

jieyouxu replied:
It's also okay to leave the PR as is, I can follow it up with refactoring the helper structs out.

jieyouxu commented:
Thank you for the fix @kxxt! The changes for the intent of this PR looks good to me. I'll make sure to follow it up with the support library work.
@bors r+ rollup

bors commented:
:pushpin: Commit 33db20978e8157f53ca24fce1c54a7b5e07159cf has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 33db20978e8157f53ca24fce1c54a7b5e07159cf -->
<!-- homu: {"type":"Approved","sha":"33db20978e8157f53ca24fce1c54a7b5e07159cf","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
@bors r-

jieyouxu commented:
@bors r=@jieyouxu,@nikic,@DianQK rollup

bors commented:
:pushpin: Commit 33db20978e8157f53ca24fce1c54a7b5e07159cf has been approved by `jieyouxu,@nikic,@DianQK`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu,@nikic,@DianQK 33db20978e8157f53ca24fce1c54a7b5e07159cf -->
<!-- homu: {"type":"Approved","sha":"33db20978e8157f53ca24fce1c54a7b5e07159cf","approver":"jieyouxu,@nikic,@DianQK","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
wtf is this parsing bors
@bors r-

jieyouxu commented:
@bors r=jieyouxu,nikic,DianQK rollup

bors commented:
:pushpin: Commit 33db20978e8157f53ca24fce1c54a7b5e07159cf has been approved by `jieyouxu,nikic,DianQK`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu,nikic,DianQK 33db20978e8157f53ca24fce1c54a7b5e07159cf -->
<!-- homu: {"type":"Approved","sha":"33db20978e8157f53ca24fce1c54a7b5e07159cf","approver":"jieyouxu,nikic,DianQK","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
